blssig/src/bdn/mod.rs

```diff
@@ -0,0 +1,75 @@
+// Copyright 2019-2024 ChainSafe Systems
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+//! BDN (Boneh-Drijvers-Neven) signature aggregation scheme, for preventing rogue public-key attacks.
+//!
+//! NOTE: currently uses standard BLS aggregation without coefficient weighting, hence returns incorrect values compared to go-f3.
+//!
+use crate::verifier::BLSError;
+use bls_signatures::{PublicKey, Serialize, Signature};
+use bls12_381::{G1Projective, G2Projective};
+use filecoin_f3_gpbft::PubKey;
+
+/// BDN aggregation context for managing signature and public key aggregation
+pub struct BDNAggregation {
+    pub_keys: Vec<PubKey>,
+}
+
+impl BDNAggregation {
+    pub fn new(pub_keys: &[PubKey]) -> Result<Self, BLSError> {
+        if pub_keys.is_empty() {
+            return Err(BLSError::EmptyPublicKeys);
+        }
+
+        Ok(Self {
+            pub_keys: pub_keys.to_vec(),
+        })
+    }
```

Comment:
Let's avoid the allocation in `.to_vec`

```suggestion
    pub fn new(pub_keys: Vec<PubKey>) -> Result<Self, BLSError> {
        if pub_keys.is_empty() {
            return Err(BLSError::EmptyPublicKeys);
        }

        Ok(Self {
            pub_keys,
        })
    }
```

Reply:
Fixed here: https://github.com/ChainSafe/rust-f3/pull/27/commits/f905f26918e175f95a3cc41d11dc3144420f78a8